Job.asyncUpdate assigns int to jobWrapper.filesToDelete

jobWrapper.filesToDelete should be a list. https://github.com/BD2KGenomics/toil/blob/master/src/toil/job.py#L602
```
#Indicate any files that should be deleted once the update of 
#the job wrapper is completed.
self.jobWrapper.filesToDelete = len(self.filesToDelete)
```
I believe this caused the testAzureSingle crash in this build: http://jenkins.cgcloud.info/job/toil-pull-requests/525/console

Okay, maybe I'm missing something. It seems like this will always leave self.jobWrapper.filesToDelete as an int if len(self.filesToDelete) is zero. And that's what happened here:
```
def clean(self, rootJobWrapper):
        """
            Function to cleanup the state of a jobStore after a restart.
            Fixes jobs that might have been partially updated.
            Resets the try counts.
            Removes jobs that are not successors of the rootJobWrapper.
            """
        #Iterate from the root jobWrapper and collate all jobs that are reachable from it
        #All other jobs returned by self.jobs() are orphaned and can be removed
        reachableFromRoot = set()
        def getConnectedJobs(jobWrapper):
            if jobWrapper.jobStoreID in reachableFromRoot:
                return
            reachableFromRoot.add(jobWrapper.jobStoreID)
            for jobs in jobWrapper.stack:
                for successorJobStoreID in map(lambda x : x[0], jobs):
                    if successorJobStoreID not in reachableFromRoot and self.exists(successorJobStoreID):
                        getConnectedJobs(self.load(successorJobStoreID))
        getConnectedJobs(rootJobWrapper)
    
        #Cleanup the state of each jobWrapper
        for jobWrapper in self.jobs():
            changed = False #Flag to indicate if we need to update the jobWrapper
            #on disk
    
>           if len(jobWrapper.filesToDelete) != 0:
E           TypeError: object of type 'int' has no len()
```
I'm not sure why there's an if statement checking whether len(self.filesToDelete) is greater than zero if that's always required to be true.

--- a/src/toil/job.py
+++ b/src/toil/job.py
@@ -375,7 +375,7 @@ class Job(object):
             """
             self.jobStore = jobStore
             self.jobWrapper = jobWrapper
-            self.localTempDir = localTempDir
+            self.localTempDir = os.path.abspath(localTempDir)
             self.loggingMessages = []
             self.filesToDelete = set()
             self.jobsToDelete = set()
@@ -445,7 +445,7 @@ class Job(object):
             handle, tmpFile = tempfile.mkstemp(prefix="tmp", 
                                                suffix=".tmp", dir=self.localTempDir)
             os.close(handle)
-            return os.path.abspath(tmpFile)
+            return os.path.abspath(tmpFile) 
 
         def writeGlobalFile(self, localFileName, cleanup=False):
             """
@@ -456,18 +456,25 @@ class Job(object):
             and all its successors have completed running. If not the global file 
             must be deleted manually.
             
-            The write is asynchronous, so further modifications during execution 
-            to the file pointed by localFileName will result in undetermined behavior. 
+            If the local file is a file returned by Job.FileStore.getLocalTempFile
+            or is in a directory, or, recursively, a subdirectory, returned by 
+            Job.FileStore.getLocalTempDir then the write is asynchronous, 
+            so further modifications during execution to the file pointed by 
+            localFileName will result in undetermined behavior. Otherwise, the 
+            method will block until the file is written to the file store. 
             """
-            jobStoreFileID = self.jobStore.getEmptyFileStoreID(None 
-                            if not cleanup else self.jobWrapper.jobStoreID)
-            self.queue.put((open(localFileName, 'r'), jobStoreFileID))
-            #Now put the file into the cache if it is a path within localTempDir
+            #Put the file into the cache if it is a path within localTempDir
             absLocalFileName = os.path.abspath(localFileName)
+            cleanupID = None if not cleanup else self.jobWrapper.jobStoreID
             if absLocalFileName.startswith(self.localTempDir):
-                #Chmod to make file read only
+                jobStoreFileID = self.jobStore.getEmptyFileStoreID(cleanupID)
+                self.queue.put((open(absLocalFileName, 'r'), jobStoreFileID))
+                #Chmod to make file read only to try to prevent accidental user modification
                 os.chmod(absLocalFileName, stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH)
                 self.jobStoreFileIDToCacheLocation[jobStoreFileID] = absLocalFileName
+            else:
+                #Write the file directly to the file store
+                jobStoreFileID = self.jobStore.writeFile(localFileName, cleanupID)
             return jobStoreFileID
         
         def writeGlobalFileStream(self, cleanup=False):
@@ -482,53 +489,71 @@ class Job(object):
             #TODO: Make this work with the caching??
             return self.jobStore.writeFileStream(None if not cleanup else self.jobWrapper.jobStoreID)
         
-        def readGlobalFile(self, fileStoreID, userPath=None):
+        def readGlobalFile(self, fileStoreID, userPath=None, cache=True):
             """
             Returns an absolute path to a local, temporary copy of the file 
             keyed by fileStoreID. 
             
-            *The returned file will be read only (have permissions 444).* 
-            
             :param userPath: a path to the name of file to which the global file will be 
-            copied or hard-linked (see below). userPath must either be: (1) a 
-            file path contained within a directory or, recursively, a subdirectory 
-            of a temporary directory returned by Job.FileStore.getLocalTempDir(), 
-            or (2) a file path returned by Job.FileStore.getLocalTempFile(). 
-            If userPath is specified and this is not true a RuntimeError exception 
-            will be raised. If userPath is specified and the file is already cached, 
+            copied or hard-linked (see below). 
+            
+            :param cache: a boolean to switch on caching (see below). Caching will 
+            attempt to keep copies of files between sequences of jobs run on the same 
+            worker. 
+            
+            If cache=True and userPath is either: 
+            (1) a file path contained within a directory or, 
+            recursively, a subdirectory of a temporary directory returned by 
+            Job.FileStore.getLocalTempDir(), or (2) a file path returned by 
+            Job.FileStore.getLocalTempFile() then the file will be cached and
+            returned file will be read only (have permissions 444).
+            If userPath is specified and the file is already cached, 
             the userPath file will be a hard link to the actual location, else it 
-            will be an actual copy of the file.  
+            will be an actual copy of the file. 
+            
+            If the cache=False or userPath is not either of the above the file 
+            will not be cached and will have default permissions. Note, if the file
+            is already cached this will result in two copies of the file on the system.
+            
+            :rtype : the absolute path to the read file
             """
             if fileStoreID in self.filesToDelete:
                 raise RuntimeError("Trying to access a file in the jobStore you've deleted: %s" % fileStoreID)
             if userPath != None:
                 userPath = os.path.abspath(userPath) #Make an absolute path
-                #Check it is a valid location
-                if not userPath.startswith(self.localTempDir):
-                    raise RuntimeError("The user path is not contained within the"
-                                       " temporary file hierarchy created by the job."
-                                       " User path: %s, temporary file root path: %s" % 
-                                       (userPath, self.localTempDir))
+                #Turn off caching if user file is not in localTempDir
+                if cache and not userPath.startswith(self.localTempDir):
+                    cache = False  
             #When requesting a new file from the jobStore first check if fileStoreID
             #is a key in jobStoreFileIDToCacheLocation.
             if fileStoreID in self.jobStoreFileIDToCacheLocation:
                 cachedAbsFilePath = self.jobStoreFileIDToCacheLocation[fileStoreID]   
-                #If the user specifies a location and it is not the current location
-                # return a hardlink to the location, else return the original location
-                if userPath == None or userPath == cachedAbsFilePath:
-                    return cachedAbsFilePath
-                if os.path.exists(userPath):
-                    os.remove(userPath)
-                os.link(cachedAbsFilePath, userPath)
-                return userPath
+                if cache:
+                    #If the user specifies a location and it is not the current location
+                    #return a hardlink to the location, else return the original location
+                    if userPath == None or userPath == cachedAbsFilePath:
+                        return cachedAbsFilePath
+                    #Chmod to make file read only
+                    if os.path.exists(userPath):
+                        os.remove(userPath)
+                    os.link(cachedAbsFilePath, userPath)
+                    return userPath
+                else:
+                    #If caching is not true then make a copy of the file
+                    localFilePath = userPath if userPath != None else self.getLocalTempFile()
+                    shutil.copyfile(cachedAbsFilePath, localFilePath)
+                    return localFilePath
             else:
                 #If it is not in the cache read it from the jobStore to the 
                 #desired location
                 localFilePath = userPath if userPath != None else self.getLocalTempFile()
                 self.jobStore.readFile(fileStoreID, localFilePath)
-                self.jobStoreFileIDToCacheLocation[fileStoreID] = localFilePath
-                #Chmod to make file read only
-                os.chmod(localFilePath, stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH)
+                #If caching is enabled and the file is in local temp dir then
+                #add to cache and make read only
+                if cache:
+                    assert localFilePath.startswith(self.localTempDir)
+                    self.jobStoreFileIDToCacheLocation[fileStoreID] = localFilePath
+                    os.chmod(localFilePath, stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH)
                 return localFilePath
 
         def readGlobalFileStream(self, fileStoreID):
@@ -601,7 +626,7 @@ class Job(object):
                     
                     #Indicate any files that should be deleted once the update of 
                     #the job wrapper is completed.
-                    self.jobWrapper.filesToDelete = len(self.filesToDelete)
+                    self.jobWrapper.filesToDelete = list(self.filesToDelete)
                     
                     #Complete the job
                     self.jobStore.update(self.jobWrapper)
@@ -638,7 +663,7 @@ class Job(object):
             At the end of the job, remove all localTempDir files except those whose value is in
             jobStoreFileIDToCacheLocation.
             
-            The param cacheSize is the total number of bytes of files allowed in the cache.
+            :param cacheSize: the total number of bytes of files allowed in the cache.
             """
             #Remove files so that the total cached files are smaller than a cacheSize
             

--- a/src/toil/jobStores/abstractJobStore.py
+++ b/src/toil/jobStores/abstractJobStore.py
@@ -134,7 +134,7 @@ class AbstractJobStore( object ):
                 for fileID in jobWrapper.filesToDelete:
                     logger.critical("Removing file in job store: %s that was marked for deletion but not previously removed" % fileID)
                     self.deleteFile(fileID)
-                jobWrapper.filesToDelete = set()
+                jobWrapper.filesToDelete = []
                 changed = True
             
             #Delete a jobWrapper if it is not reachable from the rootJobWrapper

--- a/src/toil/test/sort/sort.py
+++ b/src/toil/test/sort/sort.py
@@ -32,34 +32,44 @@ sortMemory = human2bytes('1000M')
 def setup(job, inputFile, N):
     """Sets up the sort.
     """
+    #Write the input file to the file store
+    inputFileStoreID = job.fileStore.writeGlobalFile(inputFile, True)
     job.addFollowOnJobFn(cleanup, job.addChildJobFn(down, 
-        inputFile, 0, os.path.getsize(inputFile), N).rv(), inputFile, memory=sortMemory)
+                        inputFileStoreID, N).rv(), inputFile, memory=sortMemory)
 
-def down(job, inputFile, fileStart, fileEnd, N):
+def down(job, inputFileStoreID, N):
     """Input is a file and a range into that file to sort and an output location in which
     to write the sorted file.
     If the range is larger than a threshold N the range is divided recursively and
     a follow on job is then created which merges back the results else
     the file is sorted and placed in the output.
     """
-    length = fileEnd - fileStart
+    #Read the file
+    inputFile = job.fileStore.readGlobalFile(inputFileStoreID, cache=False)
+    length = os.path.getsize(inputFile)
     if length > N:
         #We will subdivide the file
-        job.fileStore.logToMaster( "Splitting range (%i..%i) of file: %s"
-                                      % (fileStart, fileEnd, inputFile), level=logging.CRITICAL )
-        midPoint = getMidPoint(inputFile, fileStart, fileEnd)
+        job.fileStore.logToMaster( "Splitting file: %s of size: %s"
+                                      % (inputFileStoreID, length), level=logging.CRITICAL )
+        #Split the file into two copies
+        midPoint = getMidPoint(inputFile, 0, length)
+        t1 = job.fileStore.getLocalTempFile()
+        with open(t1, 'w') as fH:
+            copySubRangeOfFile(inputFile, 0, midPoint+1, fH)
+        t2 = job.fileStore.getLocalTempFile()
+        with open(t2, 'w') as fH:
+            copySubRangeOfFile(inputFile, midPoint+1, length, fH)
+        #Call down recursively
         return job.addFollowOnJobFn(up,
-            job.addChildJobFn(down, inputFile, fileStart, midPoint+1, N, memory=sortMemory).rv(),
-            job.addChildJobFn(down, inputFile, midPoint+1, fileEnd, N, memory=sortMemory).rv()).rv()          
+            job.addChildJobFn(down, job.fileStore.writeGlobalFile(t1), N, memory=sortMemory).rv(),
+            job.addChildJobFn(down, job.fileStore.writeGlobalFile(t2), N, memory=sortMemory).rv()).rv()          
     else:
         #We can sort this bit of the file
-        job.fileStore.logToMaster( "Sorting range (%i..%i) of file: %s"
-                                      % (fileStart, fileEnd, inputFile), level=logging.CRITICAL )
-        t = job.fileStore.getLocalTempFile()
-        with open(t, 'w') as fH:
-            copySubRangeOfFile(inputFile, fileStart, fileEnd, fH)
-        sort(t)
-        return job.fileStore.writeGlobalFile(t)
+        job.fileStore.logToMaster( "Sorting file: %s of size: %s"
+                                      % (inputFileStoreID, length), level=logging.CRITICAL )
+        #Sort the copy and write back to the fileStore
+        sort(inputFile)
+        return job.fileStore.writeGlobalFile(inputFile)
 
 def up(job, inputFileID1, inputFileID2):
     """Merges the two files and places them in the output.
@@ -78,9 +88,7 @@ def up(job, inputFileID1, inputFileID2):
 def cleanup(job, tempOutputFileStoreID, outputFile):
     """Copies back the temporary file to input once we've successfully sorted the temporary file.
     """
-    localTempFile = job.fileStore.readGlobalFile(tempOutputFileStoreID)
-    shutil.copyfile(localTempFile, outputFile)
-    #sort(outputFile)
+    job.fileStore.readGlobalFile(tempOutputFileStoreID, userPath=outputFile)
 
 def main():
     parser = ArgumentParser()

--- a/src/toil/test/src/jobFileStoreTest.py
+++ b/src/toil/test/src/jobFileStoreTest.py
@@ -58,22 +58,21 @@ def fileTestJob(job, inputFileStoreIDs, testStrings, chainLength):
         #Load the input jobStoreFileIDs and check that they map to the 
         #same set of random input strings
         if random.random() > 0.666:
-            #Check we get an exception if we try a file out side of the file path
-            try:
-                job.fileStore.readGlobalFile(fileStoreID, userPath="THIS/IS/NOT/A/REAL/PATH")
-            except RuntimeError:
-                pass
+            #Read the file for the fileStoreID, randomly picking a way to invoke readGlobalFile
             if random.random() > 0.5:
-                tempFile = job.fileStore.readGlobalFile(fileStoreID)
+                tempFile = job.fileStore.readGlobalFile(fileStoreID, 
+                                                        cache=random.random() > 0.5)
             else:
                 userPath = job.fileStore.getLocalTempFile()
-                tempFile = job.fileStore.readGlobalFile(fileStoreID, userPath)
+                tempFile = job.fileStore.readGlobalFile(fileStoreID, userPath, 
+                                                        cache=random.random() > 0.5)
                 assert userPath == tempFile
+            #Check the local file is as we expect
             with open(tempFile, 'r') as fH:
                 string = fH.readline()
                 assert testStrings[string[:PREFIX_LENGTH]] == string
         elif random.random() > 0.666:
-            #with job.fileStore.jobStore.readFileStream(fileStoreID) as fH:
+            #Try reading the file using the stream method.
             with job.fileStore.readGlobalFileStream(fileStoreID) as fH:
                 string = fH.readline()
                 assert testStrings[string[:PREFIX_LENGTH]] == string

